### PR TITLE
docs: update router containerization and deployment content

### PR DESCRIPTION
**Pull request by shorgi**
Updated and reorganized Containerization content, primarily rewriting Kubernetes topic as quickstarts and short examples to deploying router on k8s with Helm.

Previews: 
- [Containerization overview](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/overview)
- [Deploy in Kubernetes](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/kubernetes) 
- [Run in Docker](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker)

### CODE REVIEW

File: docs/source/containerization/kubernetes.mdx [hunk 1]
```diff
@@ -1,246 +1,274 @@
...
-router docker images.
 
-You can use helm to install charts from an OCI registry as follows:
+## Get the router Helm chart
+
+The Helm chart for each release of the Apollo Router is available from the router's GitHub repository in `<repo>/helm/chart/router`. 
+
+> For example, the Helm chart of the latest release of the router is available [here](https://github.com/apollographql/router/tree/v1.31.1/helm/chart/router).
+
+Since the router v0.14.0, Apollo has also released a Helm chart for the router as an [Open Container Initiative (OCI)](https://helm.sh/docs/topics/registries/) image in the GitHub Container registry.
```
Comment (garypen) [inline docs/source/containerization/kubernetes.mdx:41]: We prefer people to use the `OCI` release chart rather than using the version in the repo. We should probably stop documenting that altogether.

Comment (garypen) [inline docs/source/containerization/kubernetes.mdx:41]:
```suggestion
Since the router v0.14.0, Apollo has also released a Helm chart for the router as an [Open Container Initiative (OCI)](https://helm.sh/docs/topics/registries/) image in our GitHub Container registry. 
```

File: docs/source/containerization/kubernetes.mdx [hunk 2]
```diff
@@ -1,246 +1,274 @@
...
+
+The Helm chart for each release of the Apollo Router is available from the router's GitHub repository in `<repo>/helm/chart/router`. 
+
+> For example, the Helm chart of the latest release of the router is available [here](https://github.com/apollographql/router/tree/v1.31.1/helm/chart/router).
+
+Since the router v0.14.0, Apollo has also released a Helm chart for the router as an [Open Container Initiative (OCI)](https://helm.sh/docs/topics/registries/) image in the GitHub Container registry. 
+
+> The path to the OCI router chart is `oci://ghcr.io/apollographql/helm-charts/router`.
+
+<ExpansionPanel title="Click to expand example chart for router v1.31.0">
```
Comment (garypen) [inline docs/source/containerization/kubernetes.mdx:45]:
```suggestion
<ExpansionPanel title="Click to expand example values for router v1.31.0">
```

File: docs/source/containerization/kubernetes.mdx [hunk 3]
```diff
@@ -1,246 +1,274 @@
...
-```bash
-helm install --set router.configuration.telemetry.metrics.prometheus.enabled=true --set managedFederation.apiKey="REDACTED" --set managedFederation.graphRef="REDACTED" --create-namespace --namespace router-deploy router-test router --values router/values.yaml
+- A **common** file, which contains values that apply across all environments.
+- A unique **environment** file per environment, which includes and overrides the values from the common chart while adding new environment-specific values.
+
+### Common value configuration
+
+Configuring a `values.yaml` file for settings common across environments.
+
+<ExpansionPanel title="Example router Helm chart for common configuration across environments">
```
Comment (garypen) [inline docs/source/containerization/kubernetes.mdx:68]:
```suggestion
<ExpansionPanel title="Example router Helm values for common configuration across environments">
```

File: docs/source/containerization/kubernetes.mdx [hunk 4]
```diff
@@ -1,246 +1,274 @@
...
+(TODO: refine the steps between this #set-up-cluster and the next #set-up-self-hosted-router)
+
+### Set up self-hosted router
+
+Set up a self-hosted router by following the [first six steps of the self-hosted router quickstart](/graphos/quickstart/self-hosted), and stopping when you reach step [7. Deploy your router and connect clients](https://www.apollographql.com/docs/graphos/quickstart/self-hosted/#7-deploy-your-router-and-connect-clients).
+
+
+## Install router Helm chart
+
+You can install the router chart from either a local repository or the container registry.
```
Comment (garypen) [inline docs/source/containerization/kubernetes.mdx:208]:
```suggestion
You can install the router chart from either a local repository or the Apollo container registry.
```
also: Maybe repeat location here (`oci://ghcr... etc...)

File: docs/source/containerization/kubernetes.mdx [hunk 5]
```diff
@@ -1,246 +1,274 @@
...
+
+</ExpansionPanel>
+
+(TODO: explain `fullnameOverride`)
+
+Setting router command-line arguments and YAML config file keys/values. 
+
+```yaml title="values.yaml"
+router:
+  router:
```
Comment (garypen) [inline docs/source/containerization/kubernetes.mdx:80]: Everywhere where you have:
```
router:
  router:
```
The first `router:` needs to be deleted and all the rest of the text needs to be left shifted by 2 bytes.

REASON: We are using an in-house chart which "wraps" the main router chart. We do that to make it simple to deploy `Rhai` and configure `istio`. I don't think we are committing to documenting that here, since it's Apollo specific, but if we are doing that, then we need to decide how to document it.

File: docs/source/containerization/kubernetes.mdx [hunk 6]
```diff
@@ -1,246 +1,274 @@
...
+```
+
+Support for `@defer` is unsupported in gateway, so it needs to be disabled in the router.
+
+```yaml title="values.yaml"
+router:
+  router:
+    configuration:
+      supergraph:
+        defer_support: false
```
Comment (garypen) [inline docs/source/containerization/kubernetes.mdx:172]: I'm not sure if we should document this. We've enabled support for defer now.

File: docs/source/containerization/kubernetes.mdx [hunk 7]
```diff
@@ -1,246 +1,274 @@
...
+* After logging in, verify your access to the registry by showing the latest router chart from `helm`:
+
+  ```bash
+  helm show values oci://ghcr.io/apollographql/helm-charts/router
+  ```
+
+* Install the OCI image from the GiHub container registry:
+
+  ```bash title="OCI registry"
+  helm install --set router.configuration.telemetry.metrics.prometheus.enabled=true --set managedFederation.apiKey="<graph-api-key>" --set managedFederation.graphRef="<graph-ref>" --create-namespace --namespace router-deploy router-test oci://ghcr.io/apollographql/helm-charts/router --version 1.0.0-rc.8 --values router/values.yaml
```
Comment (garypen) [inline docs/source/containerization/kubernetes.mdx:223]: This needs some refinement. e.g.: prometheus we don't want to set and the multiple values files for common and env are missing.

File: docs/source/containerization/kubernetes.mdx [hunk 8]
```diff
@@ -1,246 +1,274 @@
...
+
+  (TODO: clean up and describe the used helm install flags)
+
+  ```
+  In both the following examples, we are using helm to install the router:
+  - into namespace "router-deploy" (create namespace if it doesn't exist)
+  - with helm install name "router-test"
+  - with support for prometheus enabled
+  ```
+
```
Comment (garypen) [inline docs/source/containerization/kubernetes.mdx:250]: I don't think we should document how to do this. It's better to use our OCI published chart.

**Review by garypen — COMMENTED** [review]
I've left some comments mainly for discussion when we meet later.

File: docs/source/containerization/kubernetes.mdx [hunk 9]
```diff
@@ -1,264 +1,249 @@
...
+            endpoint: <otlp-endpoint-addr>
+  ```
+
+  * `router.telemetry.metrics.prometheus` was already configured but disabled (`enabled: false`) by default. This configuration sets `enabled: true`.
+  * `router.telemetry.metrics.otlp` is enabled by inclusion.
+
+* Deploy the router with the additional YAML configuration file. For example, starting with the `helm install` command from the basic deployment step, append `--values my_values.yaml`:
+
+  ```bash
+  helm install --set managedFederation.apiKey="<graph-api-key>" --set managedFederation.graphRef="<graph-ref>"  oci://ghcr.io/apollographql/helm-charts/router --version <router-version> --values router/values.yaml --namespace <router-namespace> --values my_values.yaml
```
Comment (shorgi) [inline docs/source/containerization/kubernetes.mdx:139]: @garypen - are the steps here correct, from creating a new values YAML, editing it, and appending it to the `helm install` deployment command? (Those general steps are repeated below for Rhai scripting and the coprocessor)

Comment (garypen) [reply docs/source/containerization/kubernetes.mdx:139]: It is correct. I'd just make one cosmetic change to remove the `--namespace` entry from between the two values files. So:
```
helm install --namespace <router-namespace> --set managedFederation.apiKey="<graph-api-key>" --set managedFederation.graphRef="<graph-ref>"  oci://ghcr.io/apollographql/helm-charts/router --version <router-version> --values router/values.yaml --values my_values.yaml
```

File: docs/source/containerization/kubernetes.mdx [hunk 10]
```diff
@@ -1,264 +1,270 @@
...
 ```
 
-If you had a router running on your localhost, with default health-check configuration, you could exercise the health endpoint as follows:
+## Configure for migration from gateway
 
-curl "http://localhost:8088/health"
+When [migrating from `@apollo/gateway` to the Apollo Router](../migrating-from-gateway), consider the following tips to maximize the compatibility of your router deployment.
 
-## Using `istio` with the router
+* The gateway's maximum supported request size is 20MB, while the router's default maximum request size is less. For compatibility, set the deployed router's maximium request size to 20MB:
```
Comment (garypen) [inline docs/source/containerization/kubernetes.mdx:241]: I'm not sure about this. I think it's just that Apollo's gateway was configured to support 20MB. The router by default has a max of 2Mb, so we increased it to 20MB.

This is more general wording that will work for all users:

By default the router has a maximum supported request size of 2MB. If your gateway accepts requests larger than 2MB, you can use the following configuration to ensure that the router is compatible with your gateway deployment.

<yaml extract>

File: docs/source/containerization/kubernetes.mdx [hunk 11]
```diff
@@ -1,264 +1,270 @@
...
+* The router's timeout is increased to accommodate slow subgraphs. Only HTTP 1.1 suport is enabled.
+
+```yaml title="values.yaml"
+router:
+  configuration:
+    traffic_shaping:
+      router:
+        timeout: 6min
+      all:
+        experimental_enable_http2: false
```
Comment (garypen) [inline docs/source/containerization/kubernetes.mdx:259]: Remove line 259. We aren't using that in-house. It was a short lived experiment.

File: docs/source/containerization/kubernetes.mdx [hunk 12]
```diff
@@ -1,264 +1,270 @@
...
+      rhai:
+        scripts: <rhai-scripts-path>
+        main: <rhai-main-script>
+  ```
+
+* Deploy the router with the additional YAML configuration file. For example, starting with the `helm install` command from the basic deployment step, append `--values my_values.yaml`:
+
+  ```bash
+  helm install  --namespace <router-namespace> --set managedFederation.apiKey="<graph-api-key>" --set managedFederation.graphRef="<graph-ref>"  oci://ghcr.io/apollographql/helm-charts/router --version <router-version> --values router/values.yaml --values my_values.yaml
+  ```
```
[the remark this answers is not in the document]
Comment (garypen) [reply docs/source/containerization/kubernetes.mdx:182]: One alternative: I [already created an example repo](https://github.com/apollographql/in-house-router-example) of how to approach this problem which I have shared with customers in the past. It's a really stripped down version of how we deploy our Rhai scripts in-house.
Can we write some text in this section which explains why this is a more complex problem than the others and then links to [this repo](https://github.com/apollographql/in-house-router-example) to provide a fully worked example?

File: docs/source/containerization/kubernetes.mdx [hunk 13]
```diff
@@ -1,264 +1,270 @@
...
+
+```yaml title="my_values.yaml"
+router:
+  configuration:
+    # minimal coprocessor configuration
+    coprocessor:
+    url: http://127.0.0.1:8081 # Replace with the URL of your coprocessor's HTTP endpoint.
+    router:
+      request:
+        headers: false
```
[the remark this answers is not in the document]
Comment (shorgi) [reply docs/source/containerization/kubernetes.mdx]: Changing to show the typical coproc config

File: docs/source/containerization/kubernetes.mdx [hunk 14]
```diff
@@ -1,264 +1,270 @@
...
+  
+  * Get a GitHub OCI token and save it in an environment variable, `GITHUB_OCI_TOKEN`. For reference, follow the guide for [authenticating to the GitHub container registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-to-the-container-registry).
+
+  * Log in with the `helm registry login` command, using your saved `GITHUB_OCI_TOKEN` and GitHub username:
+
+    ```bash
+    echo ${GITHUB_OCI_TOKEN} | helm registry login -u <username> --password-stdin ghcr.io
+    ```
+
+* After logging in, verify your access to the registry by showing the latest router chart values with the `helm show values` command:
```
[the remark this answers is not in the document]
Comment (shorgi) [reply docs/source/containerization/kubernetes.mdx:87]: I needed to login (using these steps) to run `helm show values oci://ghcr.io/apollographql/helm-charts/router`, and to verify I tried it just now on my personal laptop and got a 403:

> Error: failed to authorize: failed to fetch anonymous token: unexpected status from GET request to https://ghcr.io/token?scope=repository%3Aapollographql%2Fhelm-charts%2Frouter%3Apull&scope=repository%3Auser%2Fimage%3Apull&service=ghcr.io: 403 Forbidden

Is this unexpected response from the registry?

[the remark this answers is not in the document]
Comment (garypen) [reply docs/source/containerization/kubernetes.mdx:87]: Sorry, you are correct. You do need to login. I thought they had removed that requirement, but I just checked from a machine where I knew I was definitely not logged in and confirmed I couldn't access the chart.

File: docs/shared/helm-show-router-output.mdx [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+          listen: 0.0.0.0:9090
+          path: "/metrics"
+
+  args:
+  - --hot-reload
+
+managedFederation:
+  # -- If using managed federation, the graph API key to identify router to Studio
+  apiKey:
+  # -- If using managed federation, use existing Secret which stores the graph API key instead of creating a new one.
```
Comment (Meschreiber) [inline docs/shared/helm-show-router-output.mdx:30]:
```suggestion
  # -- If using managed federation, use an existing secret which stores the graph API key instead of creating a new one.
```

File: docs/shared/router-common-config.mdx [hunk 16]
```diff
@@ -0,0 +1,210 @@
+```yaml
+# Common values. Overridden by in-house configurations.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+# Anything we set here applies to the in-house-router chart, and
```
Comment (Meschreiber) [inline docs/shared/router-common-config.mdx:6]:
```suggestion
# Anything set here applies to the in-house-router chart, and 
```

File: docs/shared/router-common-config.mdx [hunk 17]
```diff
@@ -0,0 +1,210 @@
+```yaml
+# Common values. Overridden by in-house configurations.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+# Anything we set here applies to the in-house-router chart, and 
+# all configuration that we set here is pushed to all sub-charts
```
Comment (Meschreiber) [inline docs/shared/router-common-config.mdx:7]:
```suggestion
# all configuration that set here is pushed to all sub-charts
```

File: docs/shared/router-common-config.mdx [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+# (because of the way Helm works:
```
Comment (Meschreiber) [inline docs/shared/router-common-config.mdx:8]:
```suggestion
# because of the way Helm works:
```

File: docs/shared/router-common-config.mdx [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+#  https://helm.sh/docs/chart_template_guide/subcharts_and_globals/)
```
Comment (Meschreiber) [inline docs/shared/router-common-config.mdx:9]:
```suggestion
#  https://helm.sh/docs/chart_template_guide/subcharts_and_globals/
```

File: docs/shared/router-common-config.mdx [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+        experimental_http_max_request_bytes: 20000000 #20MB
+      traffic_shaping:
+        router:
+          timeout: 6min
+        all:
+          experimental_enable_http2: false # to disable for all subgraphs
+          timeout: 5min
+      supergraph:
+        introspection: true
+        # We are intentionally DISABLING DEFER SUPPORT in the Router until
```
Comment (Meschreiber) [inline docs/shared/router-common-config.mdx:33]:
```suggestion
        #  this intentionally disables `@defer` support in the router until
```

File: docs/shared/router-common-config.mdx [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+        # we are out of our transition period from the Gateway to the Router.
```
Comment (Meschreiber) [inline docs/shared/router-common-config.mdx:34]:
```suggestion
        # until the transition from the gateway to the router is complete.
```

File: docs/shared/router-common-config.mdx [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+# all configuration that we set here is pushed to all sub-charts
+# (because of the way Helm works:
+#  https://helm.sh/docs/chart_template_guide/subcharts_and_globals/)
+
+# Set fullnameOverride here for virtualservice to use
+fullnameOverride: router
+
+router:
+  serviceAccount:
+    # We do that ourselves in permissions.yaml
```
Comment (Meschreiber) [inline docs/shared/router-common-config.mdx:16]:
```suggestion
    # The service account is created permissions.yaml
```

File: docs/shared/router-common-config.mdx [hunk 23]
```diff
@@ -0,0 +1,210 @@
...
+          experimental_enable_http2: false # to disable for all subgraphs
+          timeout: 5min
+      supergraph:
+        introspection: true
+        # We are intentionally DISABLING DEFER SUPPORT in the Router until
+        # we are out of our transition period from the Gateway to the Router.
+        defer_support: false
+        listen: 0.0.0.0:80
+        path: /graphql
+      # We keep subgraph errors on in our own environment, always.
```
Comment (Meschreiber) [inline docs/shared/router-common-config.mdx:38]: Is it worth giving the rationale for this, or just to say:

```suggestion
      # We recommend keeping subgraph errors on in your own environment, always.
```

File: docs/shared/router-common-config.mdx [hunk 24]
```diff
@@ -0,0 +1,210 @@
...
+            max_queue_size: 2048
+        metrics:
+          prometheus:
+            enabled: false
+          otlp:
+            temporality: delta
+            endpoint: "${env.NODE_IP}:4317"
+        # Enabling this both includes the `apollo-trace-id` value on the HTTP response,
+        # but also creates "Response Header" nodes in the trace that we send to Studio.
+        # This should allow us to correlate traces from Studio to logs:
```
Comment (Meschreiber) [inline docs/shared/router-common-config.mdx:104]: "should" introduces a note of uncertainty

```suggestion
        # This allows us to correlate traces from Studio to logs:
```

File: docs/shared/router-common-config.mdx [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+    tags.datadoghq.com/service: "router"
+    tags.datadoghq.com/version: "{{ .Chart.AppVersion | quote }}"
+  extraVolumeMounts:
+    - name: rhai-volume
+      mountPath: /dist/rhai
+      readOnly: true
+  extraVolumes:
+    - name: rhai-volume
+      configMap:
+        # Provide the name of the ConfigMap containing the files you want
```
Comment (Meschreiber) [inline docs/shared/router-common-config.mdx:169]:
```suggestion
        # Provide the name of the ConfigMap containing the files
```

File: docs/source/containerization/docker.mdx [hunk 26]
```diff
@@ -81,11 +113,11 @@ docker run -p 4000:4000 \
   ghcr.io/apollographql/router:<image version>-debug
 ```
 
-The router will run under the control of heaptrack. The heaptrack output will be saved to the /data directory. The output can be analyzed directly using `heaptrack_gui` or `heaptrack_print` or shared with Apollo support staff.
+The router will run under the control of heaptrack. The heaptrack output will be saved to the `/data` directory. The output can be analyzed directly using `heaptrack_gui` or `heaptrack_print` or shared with Apollo support.
 
 ## Specifying the supergraph
 
-If we don't want to use uplink to retrieve our subgraph, we can manually specify the details.
+If we don't want to use [Apollo Uplink](/federation/managed-federation/uplink/) to retrieve our subgraph, we can manually specify the details.
```
Comment (Meschreiber) [inline docs/source/containerization/docker.mdx:120]: Can we add why anyone would want to do this?

**Review by Meschreiber — COMMENTED** [review]
Great additions @garypen and @shorgi ! 👏👏👏

I've included inline copy suggestions and have these general remarks:
- Love that you've moved config files to shared content for reusability. Within the content of a page, does it generally make sense to hide them in `ExpansionPanels`? [In some cases](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/kubernetes#basic-deployment) you're already doing this, and I'm wondering if there are other places it makes sense too, e.g. https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/kubernetes#deploy-with-a-coprocessor
- Structure on most pages makes sense. On the Docker page, the intro paragraph doesn't align with section headers:

image

It would be nice to get them a little more aligned. (E.g., make [Running the debug container to investigate memory issues](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker#running-the-debug-container-to-investigate-memory-issues) an H3, add [Specifying the supergraph](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker#specifying-the-supergraph) to the intro.)

Finally, though not (yet) included on this PR, the [Passing command-line arguments to the Router binary](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker#passing-command-line-arguments-to-the-router-binary) and [Debugging your container](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker#debugging-your-container) sections could use a little copyediting love too (if time permits).

E.g. The first sentence of [Passing command-line arguments to the Router binary](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker#passing-command-line-arguments-to-the-router-binary) is a little hard to follow: "By default, the router command invoked inside the published container does run not set any of the [available command-line options](https://deploy-preview-3943--apollo-router-docs.netlify.app/configuration/overview#command-line-options)."

And for [Debugging your container](https://deploy-preview-3943--apollo-router-docs.netlify.app/containerization/docker#debugging-your-container), it would be best to  [avoid using terms like "easy" or "simply" when giving users instructions](https://developers.google.com/style/word-list#easy).

PS. Apologies for starting this review with one-off suggestions instead of incorporating them all here. Was still drinking my morning coffee! 😅

File: docs/shared/helm-show-router-output.mdx (hunk 15 above)
Comment (shorgi) [reply docs/shared/helm-show-router-output.mdx:30]: @garypen - a note to incorporate the copyedits from this PR into both the release and in-house router charts

File: docs/source/containerization/kubernetes.mdx [hunk 27]
```diff
@@ -1,264 +1,250 @@
...
+* Configure chart values to export metrics, enable Rhai scripting, and deploy a coprocessor.
+* Choose chart values that best migrate from a gateway to the router.
+
+<ElasticNotice />
 
-## Sample Kubernetes Configuration
+## About the router Helm chart
 
-> Note: The Apollo Router is made available under the Elastic License v2.0 (ELv2).  This applies to its source code and all distributions, including versions installed via Helm charts.  Read [our licensing page](https://www.apollographql.com/docs/resources/elastic-license-v2-faq/) for more details.
+[Helm](https://helm.sh) is a package manager for Kubernetes (k8s). Apollo provides an application Helm chart with each release of Apollo Router in GitHub. Since the router version 0.14.0, Apollo has released the router Helm chart as an [Open Container Initiative (OCI)](https://helm.sh/docs/topics/registries/) image in our GitHub container registry.
```
[the remark this answers is not in the document]
Comment (shorgi) [reply docs/source/containerization/kubernetes.mdx:43]: It's an unfortunate name collision, as GitHub explicitly names it its ["Container registry"](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry). We need an improvement to Apollopedia, to ignore specified instances like this.

File: docs/source/containerization/kubernetes.mdx [hunk 28]
```diff
@@ -1,264 +1,250 @@
...
- - with support for prometheus enabled
+Follow this guide to deploy the Apollo Router by using Helm to install the basic chart provided with each router release. 
 
-#### Using helm chart from our Open Container Initiative (OCI) registry
+Each router chart has a `values.yaml` file with router and deployment settings. The released, unedited file has a few explicit settings, including:
 
-Starting with release 0.14.0, each time we release the router, we'll release
-our helm chart and store it in the same OCI registry in which we store our
-router docker images.
+* Default container ports for the router's (supergraph) [HTTP server](../../configuration/overview/#listen-address), [health check endpoint](../../configuration/health-checks), and [metrics endpoint](../../configuration/metrics).
```
[the remark this answers is not in the document]
Comment (shorgi) [reply docs/source/containerization/kubernetes.mdx:55]: It's meant to be a callout to the `supergraph` key in the chart's values file. I agree though it's distracting standalone, will instead add a comment in the values file snippet that `supergraph` refers to the HTTP server.

File: docs/source/containerization/kubernetes.mdx (hunk 27 above)
[the remark this answers is not in the document]
Comment (Meschreiber) [reply docs/source/containerization/kubernetes.mdx:43]: Ah I see, I think it's referred to as "container repository" in a couple of places in different pages in this PR, so it might be worth doing a find/replace.

File: docs/source/containerization/docker.mdx (hunk 26 above)
Comment (garypen) [reply docs/source/containerization/docker.mdx:120]: This is users who don't want to subscribe to our model for automatically updating the supergraph via uplink or who don't have connectivity (air-gapped) for their environment.

File: docs/source/config.json [hunk 29]
```diff
@@ -106,8 +106,8 @@
     },
     "Containerization": {
       "Overview": "/containerization/overview",
-      "Docker": "/containerization/docker",
-      "Kubernetes": "/containerization/kubernetes"
+      "Deploy on Kubernetes": "/containerization/kubernetes",
```
[the remark this answers is not in the document]
Comment (garypen) [reply docs/source/config.json:109]: Kubernetes tends to be used for production deployments. Docker is more likely to be used by individual developers debugging/testing/etc...

File: docs/shared/router-common-config.mdx (hunk 23 above)
Comment (garypen) [reply docs/shared/router-common-config.mdx:38]: We should delete that comment. We are keeping subgraph errors on in-house to provide backwards compatibility with existing Gateway behaviour. It's not ideal, since it may leak implementation detail or PII(Personal Identifiable Information), but it's a requirement of our particular deployment.

**Comment by shorgi** [comment]
> I've included inline copy suggestions and have these general remarks:

Thanks @Meschreiber for your thoughtful and thorough review 🙏 ! I've addressed all of your suggestions and general remarks.

File: docs/source/containerization/docker.mdx [hunk 30]
```diff
@@ -123,4 +150,4 @@ Usage: build_docker_image.sh [-b [-r <repo>]] [-d] [<release>]
 		build_docker_image.sh -d v0.9.1
 ```
 
-Note: The script has to be run from the `dockerfiles/diy` directory because it makes assumptions about the relative availability of various files. The example uses [debian:bullseye-slim image](https://hub.docker.com/_/debian/) for the final image build. Feel free to modify the script to use images which better suit your own needs, but be careful if using the [-d] flag since it makes the assumption that there is a `heaptrack` package available to install.
+The script has to be run from the `dockerfiles/diy` directory because it makes assumptions about the relative availability of various files. The example uses [debian:bullseye-slim image](https://hub.docker.com/_/debian/) for the final image build. Feel free to modify the script to use images which better suit your own needs, but be careful if using the [-d] flag since it makes the assumption that there is a `heaptrack` package available to install.
```
Comment (garypen) [inline docs/source/containerization/docker.mdx:153]: You can delete that first sentence. We fixed the build script a while back so that was no longer required.

File: docs/source/containerization/kubernetes.mdx [hunk 31]
```diff
@@ -1,264 +1,250 @@
...
+
+</ExpansionPanel>
+
+### Set up Helm
+
+* Install [Helm](https://helm.sh/docs/intro/install/) **version 3.x**. The Apollo Router's Helm chart requires Helm v3.x.
+
+> ⚠️ Your Kubernetes version must be compatible with Helm v3. For details, see [Helm Version Support Policy](https://helm.sh/docs/topics/version_skew/#supported-version-skew).
+
+* Log in with Helm to the Apollo container registry in GitHub.
```
[the remark this answers is not in the document]
Comment (garypen) [reply docs/source/containerization/kubernetes.mdx:75]: Like @shorgi noted above, this is a "registry" and not a "repository". It will probably cause more confusion if we change this.

File: docs/source/containerization/docker.mdx (hunk 30 above)
Comment (shorgi) [reply docs/source/containerization/docker.mdx:153]: deleting `The script has to be run ... files.`

**Comment by garypen** [comment]
I opened a replacement PR: https://github.com/apollographql/router/pull/4001, so that we can target these changes to main. Closing this one so it doesn't get merged by mistake.
